### PR TITLE
CEDS-2681 Additional Fiscal References to add to list and 2 step remove pattern

### PR DESCRIPTION
**Pull request by janeturner**
Create AdditionalFiscalReferencesAdd page with input fields for VAT details

AdditionalFiscalReferences page:
- Remove input fields
- create table displaying VAT details with Remove link
- add Yes/No question to add more details

Create AdditionalFiscalReferencesRemove page with:
- Summary list displaying VAT details to remove
- yes/no question

### CODE REVIEW

File: conf/declaration.routes [hunk 1]
```diff
@@ -268,11 +268,17 @@ POST         /items/:itemId/fiscal-information   controllers.declaration.FiscalI
 
 # Additional Fiscal References
 
-GET         /items/:itemId/additional-fiscal-references             controllers.declaration.AdditionalFiscalReferencesController.displayPage(mode: models.Mode ?= models.Mode.Normal, itemId)
+GET         /items/:itemId/additional-fiscal-references-list        controllers.declaration.AdditionalFiscalReferencesController.displayPage(mode: models.Mode ?= models.Mode.Normal, itemId)
```
Comment (SteveSugden) [inline conf/declaration.routes:271]: Remember to let Alex Stanton know the changed/new urls

**Review by SteveSugden — DISMISSED** [review]
LGTM

File: app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html [hunk 2]
```diff
@@ -76,11 +76,11 @@
                 Seq(
                     TableRow(
                         content = Text(additionalReference.countryName),
-                        attributes = Map("id" -> s"additional_fiscal_references-row$index-label")
+                        attributes = Map("id" -> s"additional_fiscal_references-row$index-country")
```
Comment (MaciejRewera) [inline app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html:79]: Having `row$index` will result in sth like `row1`, `row2`. Are you sure this is what you want? Or maybe you'd prefer `row-1`, `row-2` ? I don't remember which way it is done in other similar places.

**Comment by MaciejRewera** [comment]
Please, please add test cases for changes in commit I provided. It is super important to cover all cases in that Controller.

File: app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html (hunk 2 above)
Comment (janeturner) [reply app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html:79]: I probably copied this from additional_information, but I can change it if the other way is better

File: app/controllers/declaration/AdditionalFiscalReferencesController.scala [hunk 3]
```diff
@@ -18,99 +18,62 @@ package controllers.declaration
...
-        navigator.continueTo(mode, routes.AdditionalFiscalReferencesController.displayPage(_, itemId))
-      case None => navigator.continueTo(mode, routes.ItemsSummaryController.displayItemsSummaryPage(_))
-    }
   }
 
-  private def badRequest(mode: Mode, itemId: String, formWithErrors: Form[AdditionalFiscalReference], references: Seq[AdditionalFiscalReference])(
-    implicit request: JourneyRequest[AnyContent]
-  ): Result =
-    BadRequest(additionalFiscalReferencesPage(mode, itemId, formWithErrors, references))
+  private def anotherYesNoForm: Form[YesNoAnswer] = YesNoAnswer.form(errorKey = "declaration.additionalFiscalReferences.add.another.empty")
```
Comment (MaciejRewera) [inline app/controllers/declaration/AdditionalFiscalReferencesController.scala:75]: Nit: The name was misleading to me. When I read it first time I though "just another Yes/No form". Maybe `addAnotherYesNoForm` would be more clear?

File: app/controllers/declaration/AdditionalFiscalReferencesRemoveController.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                  Future.successful(returnToSummary(mode, itemId))
+              }
+            }
+          )
+      case _ => Future.successful(returnToSummary(mode, itemId))
+    }
+
+  }
+
+  private def removeYesNoForm: Form[YesNoAnswer] = YesNoAnswer.form(errorKey = "declaration.additionalFiscalReferences.remove.empty")
```
Comment (MaciejRewera) [inline app/controllers/declaration/AdditionalFiscalReferencesRemoveController.scala:77]: Nit: I'd suggest changing this name to `removalYesNoForm`. The method returns the form, but the name suggests something different when you don't have a context.

File: app/controllers/declaration/AdditionalFiscalReferencesRemoveController.scala [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+  private def afterRemove(mode: Mode, itemId: String, declaration: Option[ExportsDeclaration])(implicit request: JourneyRequest[AnyContent]) =
```
Comment (MaciejRewera) [inline app/controllers/declaration/AdditionalFiscalReferencesRemoveController.scala:79]: IMHO method's name should describe WHAT the method is doing, not WHEN to use it. Can you please come up with a better name for it?

File: app/controllers/declaration/FiscalInformationController.scala [hunk 6]
```diff
@@ -52,17 +52,27 @@ class FiscalInformationController @Inject()(
...
+
+    if (mode == Mode.Change) {
+      displayFiscalInformationPage()
+    } else if (fastForward && cacheContainsFiscalReferenceData) {
+      navigator.continueTo(mode, routes.AdditionalFiscalReferencesController.displayPage(_, itemId))
+    } else if (fastForward && cacheItemIneligibleForOSR) {
+      navigator.continueTo(mode, routes.ProcedureCodesController.displayPage(_, itemId))
+    } else {
+      if (cacheContainsFiscalReferenceData) {
+        navigator.continueTo(mode, controllers.declaration.routes.AdditionalFiscalReferencesController.displayPage(_, itemId))
```
Comment (MaciejRewera) [inline app/controllers/declaration/FiscalInformationController.scala:71]: @PatrykRudnicki @SteveSugden Do you happen to know what the `fastForward` flag is for? Also, I am right to say that this case is redundant? 
It is executed if `cacheContainsFiscalReferenceData` is `true`, no matter what is the value for `fastForward`.

File: app/controllers/navigation/Navigator.scala [hunk 7]
```diff
@@ -230,14 +234,15 @@ object Navigator {
...
-    case page                          => throw new IllegalArgumentException(s"Navigator back-link route not implemented for $page on occasional")
+    case PackageInformation                => controllers.declaration.routes.NactCodeSummaryController.displayPage
+    case AdditionalInformationRequired     => controllers.declaration.routes.PackageInformationSummaryController.displayPage
+    case AdditionalInformationSummary      => controllers.declaration.routes.PackageInformationSummaryController.displayPage
+    case CusCode                           => controllers.declaration.routes.UNDangerousGoodsCodeController.displayPage
+    case NactCode                          => controllers.declaration.routes.NactCodeSummaryController.displayPage
+    case NactCodeFirst                     => controllers.declaration.routes.TaricCodeSummaryController.displayPage
+    case CommodityMeasure                  => controllers.declaration.routes.PackageInformationSummaryController.displayPage
+    case AdditionalFiscalReferencesSummary => controllers.declaration.routes.ProcedureCodesController.displayPage
```
Comment (MaciejRewera) [inline app/controllers/navigation/Navigator.scala:244]: This line seems to be the same for all journeys. Should that not be put into `commonItem` partial function? I'd expect it to be similar to `FiscalInformation`

File: app/controllers/declaration/FiscalInformationController.scala (hunk 6 above)
Comment (PatrykRudnicki) [reply app/controllers/declaration/FiscalInformationController.scala:71]: I don't remember but it was something with summary pages probably.

I think it's necessary.

The easiest way to find out what this is doing is to search in the codebase where we're setting fastForward to true

Comment (SteveSugden) [reply app/controllers/declaration/FiscalInformationController.scala:71]: I think it was something to do with making the back-button work correctly.  But I can't remember the details.  But if you break it, I think there are ATs for it.

File: app/forms/declaration/AdditionalFiscalReference.scala [hunk 8]
```diff
@@ -37,12 +42,12 @@ object AdditionalFiscalReference extends DeclarationPage {
...
       .verifying("declaration.additionalFiscalReferences.reference.empty", _.trim.nonEmpty)
       .verifying("declaration.additionalFiscalReferences.reference.error", isAlphanumeric and noLongerThan(15))
-  )(AdditionalFiscalReference.apply)(AdditionalFiscalReference.unapply)
+  )(AdditionalFiscalReference.build)(AdditionalFiscalReference.unapply)
 
   def form(): Form[AdditionalFiscalReference] = Form(mapping)
 }
 
-case class AdditionalFiscalReferencesData(references: Seq[AdditionalFiscalReference]) {
+case class AdditionalFiscalReferencesData(isRequired: Option[YesNoAnswer], references: Seq[AdditionalFiscalReference]) {
```
Comment (MaciejRewera) [inline app/forms/declaration/AdditionalFiscalReference.scala:50]: I couldn't find any usage of this field `isRequired`. Is it still necessary to have?

File: app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                        content = Text(additionalReference.reference),
+                        attributes = Map("id" -> s"additional_fiscal_references-row-$index-reference")
+                    ),
+                    TableRow(
+                        content = HtmlContent(link(
+                            messages("site.remove"),
+                            Some(messages("declaration.additionalInformation.table.remove.hint", additionalReference.reference)),
+                            AdditionalFiscalReferencesRemoveController.displayPage(mode, itemId, createId(index, additionalReference)))),
+                        classes = "govuk-table__cell--numeric",
+                        attributes = Map("id" -> s"additional_fiscal_references-row$index-remove_button")
```
Comment (MaciejRewera) [inline app/views/declaration/fiscalInformation/additional_fiscal_references.scala.html:91]: In the last `TableRow` the id is set to have `row$index` while in the others it is `row-$index`

File: test/unit/controllers/declaration/FiscalInformationControllerSpec.scala [hunk 10]
```diff
@@ -99,6 +99,18 @@ class FiscalInformationControllerSpec extends ControllerSpec with OptionValues {
         theResponseForm.value.value.onwardSupplyRelief mustBe yes
       }
 
+      "mode is Change and there is additionalFiscalReferencesData in cache" in {
```
Comment (MaciejRewera) [inline test/unit/controllers/declaration/FiscalInformationControllerSpec.scala:102]: Nit: It is enough to say "mode is Change". This way it is clear that it is not dependent on any other variable.

File: test/views/declaration/fiscalInformation/AdditionalFiscalReferencesViewSpec.scala [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        view must containElement("button").withName(SaveAndContinue.toString)
+      }
+
+      "display 'Save and return' button" in {
+        view must containElement("button").withName(SaveAndReturn.toString)
+      }
+
+      "display table header" in {
+        view.getElementsByTag("th").get(0) must containMessage("declaration.additionalFiscalReferences.country.header")
+        view.getElementsByTag("th").get(1) must containMessage("declaration.additionalFiscalReferences.numbers.header")
```
Comment (MaciejRewera) [inline test/views/declaration/fiscalInformation/AdditionalFiscalReferencesViewSpec.scala:80]: Is it possible to get these elements by class name insteead of tag?

**Review by MaciejRewera — COMMENTED** [review]
Looks good.
I am only concerned about the `Navigator` implementation. This is only why I did not approve it yet.

File: app/controllers/declaration/FiscalInformationController.scala (hunk 6 above)
Comment (MaciejRewera) [reply app/controllers/declaration/FiscalInformationController.scala:71]: It is only used once in `Navigator `(set to true):
`case CommodityDetails     => controllers.declaration.routes.FiscalInformationController.displayPage(_, _, fastForward = true)`

Comment (MaciejRewera) [reply app/controllers/declaration/FiscalInformationController.scala:71]: Although I am still not sure about the intent. Thank you for AT hint. may try it later.

File: app/controllers/navigation/Navigator.scala [hunk 12]
```diff
@@ -262,13 +262,13 @@ object Navigator {
...
   val commonItem: PartialFunction[DeclarationPage, (Mode, String) => Call] = {
-    case FiscalInformation         => controllers.declaration.routes.ProcedureCodesController.displayPage
-    case AdditionalFiscalReference => controllers.declaration.routes.FiscalInformationController.displayPage(_, _, fastForward = false)
-    case CommodityDetails          => controllers.declaration.routes.FiscalInformationController.displayPage(_, _, fastForward = true)
-    case UNDangerousGoodsCode      => controllers.declaration.routes.CommodityDetailsController.displayPage
-    case TaricCode                 => controllers.declaration.routes.TaricCodeSummaryController.displayPage
-    case TaricCodeFirst            => controllers.declaration.routes.CusCodeController.displayPage
-    case StatisticalValue          => controllers.declaration.routes.NactCodeSummaryController.displayPage
+    case FiscalInformation                 => controllers.declaration.routes.ProcedureCodesController.displayPage
+    case AdditionalFiscalReferencesSummary => controllers.declaration.routes.ProcedureCodesController.displayPage
```
Comment (MaciejRewera) [inline app/controllers/navigation/Navigator.scala:266]: 👍

**Review by MaciejRewera — APPROVED** [review]
LGTM. Good job!

**Pull request merged**